### ISSUE TITLE
Not an issue, but a question

### ISSUE THREAD

**Issue by RobertboothVentech** [stmosher/AWS_VPC_with_CSR1000v_v2.0#3]
This looks absolutely awesome!  Amazing work.  I am new to AWS in the last year and need to automate the deployment of "spoke like" vpcs with two CSRs configured for HA.  I already have our Routing/Transit VPC up and numerous "spoke like" VPCs with 2xCSRs & HA connecting back to my routing/transit vpc by way of simple GRE tunnel.   Can these scripts be used to deploy just the SPOKE portion - A single VPC with two CSRs & HA with a have the CSRs be configured with my GRE tunnel information needed for them to connect to my routing VPC like all others?

**Comment by stmosher** [comment]
Hi @RobertboothVentch,
Yes; this builds the spoke VPCs. See the template for AWS/us-east-1/A(AvailabilityZoneA) in  AWS_VPC_with_CSR1000v_v2.0/projects/DMVPN/TEMPLATES/Devices/CSR1000v/AWS/us-east-1/A/AWS_us_east_1.j2.
You can edit that jinja2 template with static GRE tunnel information to replicate exactly what you are doing now. 
The posted template configures dynamic multipoint VPN (DMVPN) with one mGRE tunnel that makes the tunneling very easy, i.e., all spokes connect via one GRE NBMA interface. I realize not everyone wants full connectivity between VPCs, so included is a whitelist security policy. You specifically permit VPC to VPC access when the VPCs are deloyed. There's also an option to modify access later, and my favorite part, when you use the program to remove a VPC, any permit entires in the other VPC CSR1000vs are removed. 
Thanks!
Steve

**Comment by RobertboothVentech** [comment]
Steve,

  I’ve been fairly stoked to get this working, but have not been successful over the last serveral days.   At the beginning I ran into several python code issues (I am running same 2.7.6) which took me a while to figure out.   They were all minor.  I have attached a copy of what I am trying to run which really doesn’t very in the main.py file; however, I am now getting the following error and can’t get back it and not sure what to do as I don’t have any file with that name and know its not a create problem.   I did confirm that my access key and everything with AWS as I am able to do things via AWS cli and or powershell….


Check the below and verify this is the connectivity you want to allow:
['cloud', 'external']

Is this what you want? Type yes or no: yes
Error asking for user input: 1 error(s) occurred:

* provider.aws: 1:3: unknown variable accessed: var.aws_region in:

${var.aws_region}
Error asking for user input: 1 error(s) occurred:

* provider.aws: 1:3: unknown variable accessed: var.aws_access_key in:

${var.aws_access_key}
Traceback (most recent call last):
  File "main.py", line 1300, in <module>
    main()
  File "main.py", line 35, in main
    addvpc()
  File "main.py", line 139, in addvpc
    with open("VPCs/{}/terraform.tfstate".format(VPC)) as json_data:
IOError: [Errno 2] No such file or directory: 'VPCs/1/terraform.tfstate'


Best regards,

[cid:image001.png@01D2E217.C102FF50]<http://www.hpe.com/>

Robert L. Booth
HCQIS Network Operations and Engineering
Technology Consultant
Hewlett-Packard Company Enterprise | U.S. Public Sector Federal Consulting
Phone: +1 571.352.5795
Robert.Booth@ventech.hcqis.org<mailto:Robert.Booth@ventech.hcqis.org>
Robert.l.booth@hpe.com<mailto:Robert.l.booth@hpe.com>
www.dxc.technology


Note: Please update your records with my new mobile phone number (571) 352.5795.  Previous number (703) 407-5806 is no longer valid.

**Comment by RobertboothVentech** [comment]
BTW, I am going to head home for the day and get an early start tomorrow morning.


Best regards,

[cid:image001.png@01D2E218.276AF860]<http://www.hpe.com/>

Robert L. Booth
HCQIS Network Operations and Engineering
Technology Consultant
Hewlett-Packard Company Enterprise | U.S. Public Sector Federal Consulting
Phone: +1 571.352.5795
Robert.Booth@ventech.hcqis.org<mailto:Robert.Booth@ventech.hcqis.org>
Robert.l.booth@hpe.com<mailto:Robert.l.booth@hpe.com>
www.dxc.technology


Note: Please update your records with my new mobile phone number (571) 352.5795.  Previous number (703) 407-5806 is no longer valid.

From: Booth, Robert
Sent: Saturday, June 10, 2017 6:33 PM
To: 'stmosher/AWS_VPC_with_CSR1000v_v2.0'; stmosher/AWS_VPC_with_CSR1000v_v2.0
Cc: Author
Subject: RE: [stmosher/AWS_VPC_with_CSR1000v_v2.0] Not an issue, but a question (#3)

Steve,

  I’ve been fairly stoked to get this working, but have not been successful over the last serveral days.   At the beginning I ran into several python code issues (I am running same 2.7.6) which took me a while to figure out.   They were all minor.  I have attached a copy of what I am trying to run which really doesn’t very in the main.py file; however, I am now getting the following error and can’t get back it and not sure what to do as I don’t have any file with that name and know its not a create problem.   I did confirm that my access key and everything with AWS as I am able to do things via AWS cli and or powershell….


Check the below and verify this is the connectivity you want to allow:
['cloud', 'external']

Is this what you want? Type yes or no: yes
Error asking for user input: 1 error(s) occurred:

* provider.aws: 1:3: unknown variable accessed: var.aws_region in:

${var.aws_region}
Error asking for user input: 1 error(s) occurred:

* provider.aws: 1:3: unknown variable accessed: var.aws_access_key in:

${var.aws_access_key}
Traceback (most recent call last):
  File "main.py", line 1300, in <module>
    main()
  File "main.py", line 35, in main
    addvpc()
  File "main.py", line 139, in addvpc
    with open("VPCs/{}/terraform.tfstate".format(VPC)) as json_data:
IOError: [Errno 2] No such file or directory: 'VPCs/1/terraform.tfstate'


Best regards,

[cid:image001.png@01D2E218.276AF860]<http://www.hpe.com/>

Robert L. Booth
HCQIS Network Operations and Engineering
Technology Consultant
Hewlett-Packard Company Enterprise | U.S. Public Sector Federal Consulting
Phone: +1 571.352.5795
Robert.Booth@ventech.hcqis.org<mailto:Robert.Booth@ventech.hcqis.org>
Robert.l.booth@hpe.com<mailto:Robert.l.booth@hpe.com>
www.dxc.technology<http://www.dxc.technology>


Note: Please update your records with my new mobile phone number (571) 352.5795.  Previous number (703) 407-5806 is no longer valid.

**Comment by stmosher** [comment]
Hi Robert,
I left key values pairs for VPC1 and VPC2 one file and VPC2, VPC2, and external in another when I uploaded them.

The solution to your problem, since you are using VPC1, is to make the below files like that from the cat output.

DMVPN/SecurityPolicies$ cat VPCandCIDRdictionary.json
{"cloud": "10.0.0.0/8"}

DMVPN/SecurityPolicies$ cat VPCandCIDRdictionary.json
{}
Note - This will be populated through creating VPCs and emptied as VPCs are removed with the program.

I'm happy to help with any issues you run into. 
Thanks,
Steve

**Comment by stmosher** [comment]
Hi Robert,
One more thing - there needs to be a "Policies" folder in the SecurityPolicies dir. Github doesn't show it since it's empty (I'll add a file in the directory in a minute...).
e.g. - DMVPN/SecurityPolicies$ mkdir Policies
Thanks,
Steve

**Comment by RobertboothVentech** [comment]
Thanks Steve,

   I took care of most of my issues; however, after launching the first CSR, I am getting the error below and not sure how to resolve it.  It says that the instance ID does not exist but it does.

Error applying plan:

1 error(s) occurred:

* aws_instance.CSR1000vA: 1 error(s) occurred:

* aws_instance.CSR1000vA: InvalidInstanceID.NotFound: The instance ID 'i-0bf45d1fbd625ccaf' does not exist
        status code: 400, request id: d161b2c1-ce2d-4d8b-8ba5-6d5780662b1a
aws_vpc.default: Refreshing state... (ID: vpc-a9c86bd0)
aws_subnet.HCQIS_XTESTING_XPORT_AZ1: Refreshing state... (ID: subnet-e5e287ad)
aws_subnet.HCQIS_XTESTING_MGT_AZ1: Refreshing state... (ID: subnet-97e386df)
aws_security_group.SG_G2_CSR1000v: Refreshing state... (ID: sg-4e876d3f)
aws_route_table.HCQIS_XTESTING_3_ZONE_AZ1: Refreshing state... (ID: rtb-03036e7b)
aws_security_group.SG_G1_CSR1000v: Refreshing state... (ID: sg-fe856f8f)
aws_subnet.HCQIS_XTESTING_APP_AZ1: Refreshing state... (ID: subnet-96e386de)
aws_internet_gateway.default: Refreshing state... (ID: igw-59a13c3f)
aws_route_table.HCQIS_XTESTING_XPORT_AZ1: Refreshing state... (ID: rtb-830469fb)
aws_instance.CSR1000vA: Refreshing state... (ID: i-0bf45d1fbd625ccaf)
aws_route_table_association.HCQIS_XTESTING_3_ZONE_AZ1: Refreshing state... (ID: rtbassoc-a9b4fdd2)
aws_route_table_association.HCQIS_XTESTING_XPORT_AZ1: Refreshing state... (ID: rtbassoc-a6b3fadd)
Traceback (most recent call last):
  File "/usr/lib/python2.7/pdb.py", line 1314, in main
    pdb._runscript(mainpyfile)
  File "/usr/lib/python2.7/pdb.py", line 1233, in _runscript
    self.run(statement)
  File "/usr/lib/python2.7/bdb.py", line 400, in run
    exec cmd in globals, locals
  File "<string>", line 1, in <module>
  File "main3.py", line 1, in <module>
    import simplejson as json
  File "main3.py", line 34, in main
    addvpc()
  File "main3.py", line 141, in addvpc
    IP_A = d["modules"][0]["resources"]["aws_eip.CSR1000vA"]["primary"]["attributes"]["public_ip"]
KeyError: 'aws_eip.CSR1000vA'
Uncaught exception. Entering post mortem debugging
Running 'cont' or 'step' will restart the program
> /home/ubuntu/projects/DMVPN/main3.py(141)addvpc()
-> IP_A = d["modules"][0]["resources"]["aws_eip.CSR1000vA"]["primary"]["attributes"]["public_ip"]
(Pdb) Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/usr/lib/python2.7/pdb.py", line 1338, in <module>
    pdb.main()
  File "/usr/lib/python2.7/pdb.py", line 1330, in main
    pdb.interaction(None, t)
  File "/usr/lib/python2.7/pdb.py", line 210, in interaction
    self.cmdloop()
  File "/usr/lib/python2.7/cmd.py", line 130, in cmdloop
    line = raw_input(self.prompt)

Best regards,

[cid:image001.png@01D2E43C.1B308120]<http://www.hpe.com/>

Robert L. Booth
HCQIS Network Operations and Engineering
Technology Consultant
Hewlett-Packard Company Enterprise | U.S. Public Sector Federal Consulting
Phone: +1 571.352.5795
Robert.Booth@ventech.hcqis.org<mailto:Robert.Booth@ventech.hcqis.org>
Robert.l.booth@hpe.com<mailto:Robert.l.booth@hpe.com>
www.dxc.technology


Note: Please update your records with my new mobile phone number (571) 352.5795.  Previous number (703) 407-5806 is no longer valid.

**Comment by stmosher** [comment]
Hi Robert,
What version of terraform are you using? -I'll try to get the same problem
Thanks,
Steve

**Comment by RobertboothVentech** [comment]
Terraform v0.9.8


Best regards,

[cid:image001.png@01D2E44C.1649ABE0]<http://www.hpe.com/>

Robert L. Booth
HCQIS Network Operations and Engineering
Technology Consultant
Hewlett-Packard Company Enterprise | U.S. Public Sector Federal Consulting
Phone: +1 571.352.5795
Robert.Booth@ventech.hcqis.org<mailto:Robert.Booth@ventech.hcqis.org>
Robert.l.booth@hpe.com<mailto:Robert.l.booth@hpe.com>
www.dxc.technology


Note: Please update your records with my new mobile phone number (571) 352.5795.  Previous number (703) 407-5806 is no longer valid.

**Comment by RobertboothVentech** [comment]
For whatever reason, which I am still searching, it is failing at the modifyinstanceattribute…. The CSR launches and the instanceid is correct as I can see terraform talking to it earlier in the sript.

2017/06/13 17:55:41 [DEBUG] plugin: terraform: Action=ModifyInstanceAttribute&InstanceId=i-0f8fb793ea4f3885b&SourceDestCheck.Value=false&Version=2016-11-15
2017/06/13 17:55:41 [DEBUG] plugin: terraform: -----------------------------------------------------
2017/06/13 17:55:41 [DEBUG] root: eval: *terraform.EvalWriteState
2017/06/13 17:55:41 [DEBUG] root: eval: *terraform.EvalApplyProvisioners
2017/06/13 17:55:41 [DEBUG] root: eval: *terraform.EvalIf
2017/06/13 17:55:41 [DEBUG] root: eval: *terraform.EvalWriteState
2017/06/13 17:55:41 [DEBUG] root: eval: *terraform.EvalWriteDiff
2017/06/13 17:55:41 [DEBUG] root: eval: *terraform.EvalApplyPost
2017/06/13 17:55:41 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* aws_instance.CSR1000vA: InvalidInstanceID.NotFound: The instance ID 'i-0f8fb793ea4f3885b' does not exist
        status code: 400, request id: 3ac1a0eb-9feb-493f-8114-9498d7e7b59a
2017/06/13 17:55:41 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* aws_instance.CSR1000vA: InvalidInstanceID.NotFound: The instance ID 'i-0f8fb793ea4f3885b' does not exist
        status code: 400, request id: 3ac1a0eb-9feb-493f-8114-9498d7e7b59a


Best regards,

[cid:image001.png@01D2E450.7CB18110]<http://www.hpe.com/>

Robert L. Booth
HCQIS Network Operations and Engineering
Technology Consultant
Hewlett-Packard Company Enterprise | U.S. Public Sector Federal Consulting
Phone: +1 571.352.5795
Robert.Booth@ventech.hcqis.org<mailto:Robert.Booth@ventech.hcqis.org>
Robert.l.booth@hpe.com<mailto:Robert.l.booth@hpe.com>
www.dxc.technology


Note: Please update your records with my new mobile phone number (571) 352.5795.  Previous number (703) 407-5806 is no longer valid.

**Comment by RobertboothVentech** [comment]
I wonder if I am running into eventual consistency….
http://docs.aws.amazon.com/AWSEC2/latest/APIReference/query-api-troubleshooting.html#eventual-consistency


Best regards,

[cid:image001.png@01D2E455.EC97F360]<http://www.hpe.com/>

Robert L. Booth
HCQIS Network Operations and Engineering
Technology Consultant
Hewlett-Packard Company Enterprise | U.S. Public Sector Federal Consulting
Phone: +1 571.352.5795
Robert.Booth@ventech.hcqis.org<mailto:Robert.Booth@ventech.hcqis.org>
Robert.l.booth@hpe.com<mailto:Robert.l.booth@hpe.com>
www.dxc.technology


Note: Please update your records with my new mobile phone number (571) 352.5795.  Previous number (703) 407-5806 is no longer valid.

**Comment by stmosher** [comment]
Could be.. I ran into something like a few time today in Azure today where terraform would throw an error but I could then go into the VPC folder, run terraform again, and it would work.

I'm changing the subprocess terraform apply commands to use the below:
def apply_terraform(VPC):
    exitcode = 1
    while exitcode != 0:
        print "Applying Terraform Cloud Definition"
        cmd = "terraform apply"
        args = shlex.split(cmd)
        proc = Popen(args, stdout=PIPE, stderr=PIPE, cwd="VPCs/{}".format(VPC))
        out, err = proc.communicate()
        print out
        print err
        exitcode = proc.returncode
    return

I've got a few more updates such as when the CSR1000v responds to SSH with a banner but isn't ready to accept ansible configuration (I added an ansible show running-config that must work before router configuration).

I'll email you my lastest once it's updated.
Best Regards,
Steve
